### ISSUE TITLE
No error when type annotation for `self.foo` comes after type is already inferred

### ISSUE THREAD

**Issue by gvanrossum** [python/mypy#6571]
(Not sure if this is a dupe.)

This is not an error:
```
class C:
    def __init__(self) -> None:
        self.foo = 12
        self.foo: int = 12  # Expected: self.foo is already defined on line 4
```
Even though this is:
```
foo = 12
foo: int = 12   # E: Name 'foo' already defined on line 1
```
It's the same if the annotation has the form of a `# type:` comment.

It's particularly obnoxious if the inferred type is wider than the type annotation, e.g. `Any` -- the inferred type silently wins.

**Comment by ilevkivskyi** [comment]
Note this also happens with the new semantic analyzer.

**Comment by aaalex5** [comment]
I'm having reproducing the bug. I am using the code snippets that you suggest with the command "mypy test.py", where test.py includes your second snippet, but without any error. Could you give more information on how to reproduce the bug?

**Comment by aaalex5** [comment]
I would like to work on this issue, but I am unable to reproduce the bug. Any help on how I may be able to reproduce this?

**Comment by ilevkivskyi** [comment]
I can reproduce this false negative. Just copy it to a file and type-check it. Also I don't think it is a good issue for beginners.
